framework/src/variables/MooseVariableData.C

```diff
@@ -424,537 +430,201 @@ MooseVariableData<OutputType>::computeValues()
-    if (_need_div)
-      _div_u[i] = 0;
-
-    if (_need_second_previous_nl)
-      _second_u_previous_nl[i] = 0;
-
-    if (is_transient)
+    else
     {
-      if (_need_u_dot)
-        _u_dot[i] = 0;
-
-      if (_need_u_dotdot)
-        _u_dotdot[i] = 0;
-
-      if (_need_u_dot_old)
-        _u_dot_old[i] = 0;
-
-      if (_need_u_dotdot_old)
-        _u_dotdot_old[i] = 0;
-
-      if (_need_du_dot_du)
-        _du_dot_du[i] = 0;
-
-      if (_need_du_dotdot_du)
-        _du_dotdot_du[i] = 0;
-
-      if (_need_grad_dot)
-        _grad_u_dot[i] = 0;
-
-      if (_need_grad_dotdot)
-        _grad_u_dotdot[i] = 0;
-
-      if (_need_second_old)
-        _second_u_old[i] = 0;
-
-      if (_need_second_older)
-        _second_u_older[i] = 0;
-
-      if (_need_curl_old)
-        _curl_u_old[i] = 0;
-
-      if (_need_div_old)
-        _div_u_old[i] = 0;
+      _mapped_grad_phi_face.resize(num_dofs);
+      for (unsigned int i = 0; i < num_dofs; i++)
+      {
+        _mapped_grad_phi_face[i].resize(nqp, Eigen::Map<RealDIMValue>(nullptr));
+        for (unsigned int qp = 0; qp < nqp; qp++)
+          // Note: this does NOT do any allocation.  It is "reconstructing" the object in place
+          new (&_mapped_grad_phi_face[i][qp])
+              Eigen::Map<RealDIMValue>(const_cast<Real *>(&(*_current_grad_phi)[i][qp](0)));
+      }
     }
   }
 
-  bool second_required =
-      _need_second || _need_second_old || _need_second_older || _need_second_previous_nl;
-  bool curl_required = _need_curl || _need_curl_old;
-  bool div_required = _need_div || _need_div_old;
+  mooseAssert(
+      !(_need_second || _need_second_old || _need_second_older || _need_second_previous_nl) ||
+          _current_second_phi,
+      "We're requiring a second calculation but have not set a second shape function!");
+  mooseAssert(!(_need_curl || _need_curl_old) || _current_curl_phi,
+              "We're requiring a curl calculation but have not set a curl shape function!");
+  mooseAssert(!(_need_div || _need_div_old) || _current_div_phi,
+              "We're requiring a divergence calculation but have not set a div shape function!");
 
-  for (unsigned int i = 0; i < num_dofs; i++)
+  const auto fill = [this, &nqp, &num_dofs](auto & dest, const auto & phi, const auto & vector)
   {
-    for (unsigned int qp = 0; qp < nqp; qp++)
+    constexpr bool is_real = std::is_same_v<OutputType, Real>;
+    constexpr bool is_real_vector = std::is_same_v<OutputType, RealVectorValue>;
+    constexpr bool is_eigen = std::is_same_v<OutputType, RealEigenVector>;
+    mooseAssert(is_real || is_real_vector || is_eigen, "Unsupported type");
+
+    if constexpr (!is_eigen)
+      (void)this;
+
+    // Output type
+    using T = typename std::remove_reference_t<decltype(dest)>::value_type;
+    constexpr bool is_output = std::is_same_v<T, OutputType>;
+    constexpr bool is_gradient = std::is_same_v<T, OutputGradient>;
+    constexpr bool is_second = std::is_same_v<T, OutputSecond>;
+
```

Comment:
should there be a ` is_div = is_same<T, lower rank(T) >` ?
the output type of div(vector) is going to be Real